src/metabase/pulse.clj

```diff
@@ -80,29 +83,73 @@
     (compare (:row dashcard-1) (:row dashcard-2))
     (compare (:col dashcard-1) (:col dashcard-2))))
 
-(defn- virtual-card-of-type?
-  "Check if dashcard is a virtual with type `ttype`.
-
-  There are currently 3 types of virtual card: text, action, link."
-  [dashcard ttype]
-  (and (map? dashcard)
-       (= ttype (get-in dashcard [:visualization_settings :virtual_card :display]))))
+(defn- link-card->url
+  [{:keys [entity url] :as link-card}]
+  (if url
+    (:url link-card)
+    (let [{:keys [db_id id model]} entity]
+      (case model
+        "card"       (urls/card-url id)
+        "dataset"    (urls/card-url id)
+        "collection" (urls/collection-url id)
+        "dashboard"  (urls/dashboard-url id)
+        "database"   (urls/database-url id)
+        "table"      (urls/table-url db_id id)))))
+
+(defn- link-card->content
+  [{:keys [entity url] :as _link-card}]
+  (if url
+    url
+    (:name entity)))
+
+(defn- link-card->description
+  [{:keys [entity url] :as _link-card}]
+  (if url
+    nil
+    (:description entity)))
+
+(defn- link-card->text
+  [link-card]
+  {:text (str (format
+                "### [%s](%s)"
+                (link-card->content link-card)
+                (link-card->url link-card))
+              (when-let [description (link-card->description link-card)]
+                (format "\n%s" description)))})
+
+(defn- dashcard-link-card->content
+  "Convert a dashcard that is a linkcard to pulse content."
+  [{viz-settings :visualization_settings :as _dashcard}]
+  (let [link-card (:link viz-settings)]
+    (cond
+      (some? (:url link-card))
+      (link-card->text link-card)
+      ;; if link card link to an entity, update the settins because
+      ;; its stored info might be out-of-date
+      (some? (:entity link-card))
+      (let [{:keys [model id]} (:entity link-card)
+            instance           (t2/select-one
+                                 (serdes.util/link-card-model->toucan-model model)
+                                 (dashboard-card/link-card-info-query-for-model model id))]
```

Comment:
I don't think this is indirection tbh, on UI we use the same method to get the data, and we should reuse the same thing when rendering subscription.

The query for each model is sometimes different between model, but the output of it always has a consistent schema.
For example you can see in here: https://github.com/metabase/metabase/blob/b033a734d3f61ecf4362e0314cbff0d0f6e3fe5f/src/metabase/models/dashboard_card.clj#L241, for table instead of taking using the `table.name`, we use `table.display_name` and alias as `name`.

This sort of thing make reusing the query make sense to me.

Reply:
That's true. The indirection is needed for `display_name` -> `name`. All good then